rkt/api_service.go

```diff
@@ -305,76 +325,190 @@ func getNetworks(p *pod) []*v1alpha.Network {
+	}
+	if len(agg) == 1 {
+		return agg[0].Error()
+	}
+	result := fmt.Sprintf("[%s", agg[0].Error())
+	for i := 1; i < len(agg); i++ {
+		result += fmt.Sprintf(", %s", agg[i].Error())
+	}
+	result += "]"
+	return result
+}
+
+func fillStaticAppInfo(store *store.Store, p *pod, v1pod *v1alpha.Pod) error {
+	var errlist []error
+
+	// Fill static app image info.
+	for _, app := range v1pod.Apps {
+		// Fill app's image info (id, name, version).
+		fullImageID, err := store.ResolveKey(app.Image.Id)
+		if err != nil {
+			stderr.PrintE(fmt.Sprintf("failed to resolve the image ID %q", app.Image.Id), err)
+			errlist = append(errlist, err)
+		}
+
+		// The following information is always known
+		app.Image = &v1alpha.Image{
+			BaseFormat: &v1alpha.ImageFormat{
+				// Only support appc image now. If it's a docker image, then it
+				// will be transformed to appc before storing in the disk store.
+				Type:    v1alpha.ImageType_IMAGE_TYPE_APPC,
+				Version: schema.AppContainerVersion.String(),
+			},
+			Id: fullImageID,
+			// Other information are not available because they require the image
+			// info from store. Some of it is filled in below if possible.
+		}
+
+		im, err := p.getAppImageManifest(*types.MustACName(app.Name))
+		if err != nil {
+			stderr.PrintE(fmt.Sprintf("failed to get image manifests for app %q", app.Name), err)
+			errlist = append(errlist, err)
+		} else {
+			app.Image.Name = im.Name.String()
+
+			version, ok := im.Labels.Get("version")
+			if !ok {
+				version = "latest"
+			}
+			app.Image.Version = version
+		}
+	}
+
+	if len(errlist) != 0 {
+		return aggregate(errlist)
+	}
+	return nil
+}
+
+func (s *v1AlphaAPIServer) getBasicPodFromDisk(p *pod) (*v1alpha.Pod, []byte, error) {
 	pod := &v1alpha.Pod{Id: p.uuid.String(), Pid: -1}
 
-	manifest, data, err := getPodManifest(p)
+	manifest, data, err := s.getPodManifest(p)
 	if err != nil {
 		stderr.PrintE(fmt.Sprintf("failed to get the pod manifest for pod %q", p.uuid), err)
 	} else {
-		pod.Manifest = data
 		pod.Annotations = convertAnnotationsToKeyValue(manifest.Annotations)
 		pod.Apps = getApplist(manifest)
+		err = fillStaticAppInfo(s.store, p, pod)
+	}
+
+	return pod, data, err
+}
+
+// getBasicPod returns *v1alpha.Pod with basic pod information and the marshaled manifest.
+func (s *v1AlphaAPIServer) getBasicPod(p *pod) (*v1alpha.Pod, []byte) {
+	mtime, err := p.getModTime("pod")
+	if err != nil {
+		stderr.PrintE(fmt.Sprintf("failed to read the pod manifest's mtime for pod %q", p.uuid), err)
+	}
+
+	// Couldn't use pod.uuid directly as it's a pointer.
+	item, found := s.podCache.Get(p.uuid.String())
+	if found && err == nil {
+		itemValue := item.(*podCacheItem)
+
+		// Check the mtime to make sure we are not returning stale manifests.
+		if !mtime.After(itemValue.mtime) {
+			return itemValue.pod, itemValue.manifestData
+		}
+	}
+
+	pod, data, err := s.getBasicPodFromDisk(p)
+	if err == nil {
+		s.podCache.Add(p.uuid.String(), &podCacheItem{pod, data, mtime})
```

Comment:
mtime could be zero here if the `getModTime` call returns such; I think that's a special case and we neither want to hit nor write to the cache in that case because a failure to read pod doesn't seem like it gives us enough information to make a meaningful cache decision.